### ISSUE TITLE
Investigate test_truncated_gz failure

### ISSUE THREAD

**Issue by marcelm** [marcelm/xopen#16]
Possibly the simplification of `PipedGzipReader.read` introduced in #14 sometimes causes a test failure, as [seen on Travis](https://travis-ci.org/marcelm/xopen/builds/571788800).

**Comment by rhpvorderman** [comment]
Hmmm. I can run this through hyperfine with a 100 runs. Using conda I can test multiple interpreters.
If that helps.

**Comment by marcelm** [comment]
I can now actually reproduce this locally with `while tox; do true; done`. This failed after a minute or so.

**Comment by marcelm** [comment]
I think it’s a timing issue in `PipedGzipReader.close()`.

**Comment by rhpvorderman** [comment]
I cannot reproduce the error on my machine. Which makes me suspect that it is a race condition, and this machine is just to fast to get it.
I made a small fix in #17 which should prevent the race condition from happening.

**Comment by marcelm** [comment]
I also thought that adding a `.wait()` would be an idea, but there’s a [big warning in the docs](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait) that it can deadlock when used with `stdout=PIPE`, which we do use. The intention of the previous code within `read()` and `readline()` was to trigger the `.wait()` only when we were sure that we were already at EOF, which would imply that the process won’t write any further data into the kernel buffer.

I’m not sure that a `self.process.terminate()` followed by an immediate `self.process.wait()` cannot deadlock. When you read just a couple of bytes from a file and then close it before reaching the end, perhaps the signal gets queued and delivered only after the process has managed to fill the buffer already.

However, I don’t see this happening in my tests and the PR fixes this particular issue.

**Comment by rhpvorderman** [comment]
> I also thought that adding a `.wait()` would be an idea, but there’s a [big warning in the docs](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait) that it can deadlock when used with `stdout=PIPE`, which we do use. The intention of the previous code within `read()` and `readline()` was to trigger the `.wait()` only when we were sure that we were already at EOF, which would imply that the process won’t write any further data into the kernel buffer.

When working on [pytest-workflow](https://github.com/lumc/pytest-workflow) I have experienced issues with `.wait()` because I was piping stdout and sterr of a workflow engine to a PIPE and the process hanged in production because long pipelines produce more stdout than 64kb. 

Apperently Popen.communicate() can fix this. I used it successfullty with pytest-workflow. 
https://docs.python.org/3.5/library/subprocess.html#subprocess.Popen.communicate

> I’m not sure that a self.process.terminate() followed by an immediate self.process.wait() cannot deadlock. When you read just a couple of bytes from a file and then close it before reaching the end, perhaps the signal gets queued and delivered only after the process has managed to fill the buffer already.

I think `.wait()` should not cause any problems here. There are two situations when the file is closed:
1. Everything finished succesfully. The process has an exit code. No problems there.
2. The (python) program crashes and pigz is still running. Process.terminate will run in this case. Even if the kernel buffer is full the process gets a signal and it will terminate. If the buffer is full it is just waiting for it the buffer to empty, but that should not block process signals. 

So far the theory. I guess time will tell if it is correct...

**Comment by marcelm** [comment]
`Popen.communicate()` reads all the data at once. Because we want to stream in xopen, it doesn’t help here.

Regarding wait + terminate, I forgot that if a process (pigz) is blocked on a write(), it will still receive signals. So it should work.

**Comment by marcelm** [comment]
Note that I just pushed bfc714d, which fixes another issue that I saw when running the tests repeatedly. I’ll tag another release shortly.

**Comment by marcelm** [comment]
Revisiting this because of https://github.com/pycompression/xopen/pull/63#issuecomment-837644218: My comment about `communicate()` wasn’t thought out – I understand that we should rewrite everything to use `communicate()`, but using `communicate()` just before shutdown should be fine and solve this problem.
